File: src/src/main/resources/mmlspark/RecommendationIndexer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class RecommendationIndexer(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """


    Args:

        itemInputCol (str): Item Input Col
        itemOutputCol (str): Item Output Col
        ratingCol (str): Rating Col
        userInputCol (str): User Input Col
        userOutputCol (str): User Output Col
    """

    @keyword_only
    def __init__(self, itemInputCol=None, itemOutputCol=None, ratingCol=None, userInputCol=None, userOutputCol=None):
        super(RecommendationIndexer, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.RecommendationIndexer")
        self.itemInputCol = Param(self, "itemInputCol", "itemInputCol: Item Input Col")
        self.itemOutputCol = Param(self, "itemOutputCol", "itemOutputCol: Item Output Col")
        self.ratingCol = Param(self, "ratingCol", "ratingCol: Rating Col")
        self.userInputCol = Param(self, "userInputCol", "userInputCol: User Input Col")
        self.userOutputCol = Param(self, "userOutputCol", "userOutputCol: User Output Col")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, itemInputCol=None, itemOutputCol=None, ratingCol=None, userInputCol=None, userOutputCol=None):
        """
        Set the (keyword only) parameters

        Args:

            itemInputCol (str): Item Input Col
            itemOutputCol (str): Item Output Col
            ratingCol (str): Rating Col
            userInputCol (str): User Input Col
            userOutputCol (str): User Output Col
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setItemInputCol(self, value):
        """

        Args:

            itemInputCol (str): Item Input Col

        """
        self._set(itemInputCol=value)
        return self


    def getItemInputCol(self):
        """

        Returns:

            str: Item Input Col
        """
        return self.getOrDefault(self.itemInputCol)


    def setItemOutputCol(self, value):
        """

        Args:

            itemOutputCol (str): Item Output Col

        """
        self._set(itemOutputCol=value)
        return self


    def getItemOutputCol(self):
        """

        Returns:

            str: Item Output Col
        """
        return self.getOrDefault(self.itemOutputCol)


    def setRatingCol(self, value):
        """

        Args:

            ratingCol (str): Rating Col

        """
        self._set(ratingCol=value)
        return self


    def getRatingCol(self):
        """

        Returns:

            str: Rating Col
        """
        return self.getOrDefault(self.ratingCol)


    def setUserInputCol(self, value):
        """

        Args:

            userInputCol (str): User Input Col

        """
        self._set(userInputCol=value)
        return self


    def getUserInputCol(self):
        """

        Returns:

            str: User Input Col
        """
        return self.getOrDefault(self.userInputCol)


    def setUserOutputCol(self, value):
        """

        Args:

            userOutputCol (str): User Output Col

        """
        self._set(userOutputCol=value)
        return self


    def getUserOutputCol(self):
        """

        Returns:

            str: User Output Col
        """
        return self.getOrDefault(self.userOutputCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.RecommendationIndexer"

    @staticmethod
    def _from_java(java_stage):
        module_name=RecommendationIndexer.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".RecommendationIndexer"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return RecommendationIndexerModel(java_model)


class RecommendationIndexerModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`RecommendationIndexer`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.RecommendationIndexerModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=RecommendationIndexerModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".RecommendationIndexerModel"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/RecommendationIndexer.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;

using SynapseML.Dotnet.Utils;
using Synapse.ML.Recommendation;

namespace Synapse.ML.Recommendation
{
    /// <summary>
    /// <see cref="RecommendationIndexer"/> implements RecommendationIndexer
    /// </summary>
    public class RecommendationIndexer : JavaEstimator<RecommendationIndexerModel>, IJavaMLWritable, IJavaMLReadable<RecommendationIndexer>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.recommendation.RecommendationIndexer";

        /// <summary>
        /// Creates a <see cref="RecommendationIndexer"/> without any parameters.
        /// </summary>
        public RecommendationIndexer() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="RecommendationIndexer"/> with a UID that is used to give the
        /// <see cref="RecommendationIndexer"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public RecommendationIndexer(string uid) : base(s_className, uid)
        {
        }

        internal RecommendationIndexer(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets itemInputCol value for <see cref="itemInputCol"/>
        /// </summary>
        /// <param name="itemInputCol">
        /// Item Input Col
        /// </param>
        /// <returns> New RecommendationIndexer object </returns>
        public RecommendationIndexer SetItemInputCol(string value) =>
            WrapAsRecommendationIndexer(Reference.Invoke("setItemInputCol", (object)value));
        
        /// <summary>
        /// Sets itemOutputCol value for <see cref="itemOutputCol"/>
        /// </summary>
        /// <param name="itemOutputCol">
        /// Item Output Col
        /// </param>
        /// <returns> New RecommendationIndexer object </returns>
        public RecommendationIndexer SetItemOutputCol(string value) =>
            WrapAsRecommendationIndexer(Reference.Invoke("setItemOutputCol", (object)value));
        
        /// <summary>
        /// Sets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <param name="ratingCol">
        /// Rating Col
        /// </param>
        /// <returns> New RecommendationIndexer object </returns>
        public RecommendationIndexer SetRatingCol(string value) =>
            WrapAsRecommendationIndexer(Reference.Invoke("setRatingCol", (object)value));
        
        /// <summary>
        /// Sets userInputCol value for <see cref="userInputCol"/>
        /// </summary>
        /// <param name="userInputCol">
        /// User Input Col
        /// </param>
        /// <returns> New RecommendationIndexer object </returns>
        public RecommendationIndexer SetUserInputCol(string value) =>
            WrapAsRecommendationIndexer(Reference.Invoke("setUserInputCol", (object)value));
        
        /// <summary>
        /// Sets userOutputCol value for <see cref="userOutputCol"/>
        /// </summary>
        /// <param name="userOutputCol">
        /// User Output Col
        /// </param>
        /// <returns> New RecommendationIndexer object </returns>
        public RecommendationIndexer SetUserOutputCol(string value) =>
            WrapAsRecommendationIndexer(Reference.Invoke("setUserOutputCol", (object)value));

        
        /// <summary>
        /// Gets itemInputCol value for <see cref="itemInputCol"/>
        /// </summary>
        /// <returns>
        /// itemInputCol: Item Input Col
        /// </returns>
        public string GetItemInputCol() =>
            (string)Reference.Invoke("getItemInputCol");
        
        
        /// <summary>
        /// Gets itemOutputCol value for <see cref="itemOutputCol"/>
        /// </summary>
        /// <returns>
        /// itemOutputCol: Item Output Col
        /// </returns>
        public string GetItemOutputCol() =>
            (string)Reference.Invoke("getItemOutputCol");
        
        
        /// <summary>
        /// Gets ratingCol value for <see cref="ratingCol"/>
        /// </summary>
        /// <returns>
        /// ratingCol: Rating Col
        /// </returns>
        public string GetRatingCol() =>
            (string)Reference.Invoke("getRatingCol");
        
        
        /// <summary>
        /// Gets userInputCol value for <see cref="userInputCol"/>
        /// </summary>
        /// <returns>
        /// userInputCol: User Input Col
        /// </returns>
        public string GetUserInputCol() =>
            (string)Reference.Invoke("getUserInputCol");
        
        
        /// <summary>
        /// Gets userOutputCol value for <see cref="userOutputCol"/>
        /// </summary>
        /// <returns>
        /// userOutputCol: User Output Col
        /// </returns>
        public string GetUserOutputCol() =>
            (string)Reference.Invoke("getUserOutputCol");

        /// <summary>Fits a model to the input data.</summary>
        /// <param name="dataset">The <see cref="DataFrame"/> to fit the model to.</param>
        /// <returns><see cref="RecommendationIndexerModel"/></returns>
        override public RecommendationIndexerModel Fit(DataFrame dataset) =>
            new RecommendationIndexerModel(
                (JvmObjectReference)Reference.Invoke("fit", dataset));

        /// <summary>
        /// Loads the <see cref="RecommendationIndexer"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="RecommendationIndexer"/> was saved to</param>
        /// <returns>New <see cref="RecommendationIndexer"/> object, loaded from path.</returns>
        public static RecommendationIndexer Load(string path) => WrapAsRecommendationIndexer(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<RecommendationIndexer> Read() =>
            new JavaMLReader<RecommendationIndexer>((JvmObjectReference)Reference.Invoke("read"));

        private static RecommendationIndexer WrapAsRecommendationIndexer(object obj) =>
            new RecommendationIndexer((JvmObjectReference)obj);

        
    }
}

        